# FACET (Fungal Alignment Culling and Evaluation Tool)


### Download
See the [latest release](https://github.com/aast242/FACET/releases) for the most recent, fully-tested version of FACET.

# FACET Dependencies
### *BLASTn*
FACET runs off a BLAST base for most of its modules (with the exception of the outfile module). As such, it requires BLAST+ to be installed and reachable in the PATH.

Using `apt-get install ncbi-blast+` installs an older version, so BLAST+ should be installed from [NCBI's website](https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/)


### *BioPython*
FACET uses BioPython for interpereting FASTA files. BioPython can be found [here](https://biopython.org/wiki/Download) and can also be easily installed by using the command:

`pip install biopython`

If you already have BioPython installed it can be updated to the latest version by using the command:

`pip install biopython --upgrade`


### *Samtools*
FACET uses samtools to convert SAM files to BAM files and generate BAM index files (.bai). The github page for samtools can be found [here](https://github.com/samtools/samtools) and the software can also be easily installed by running the following command:

`apt-get install samtools`

# What is FACET?
FACET is a toolkit for removing redundancy in BLAST searches for repetitive elements in fungal genomes.
FACET can also be used for other applications like **genome masking**, **identifing repetitive regions of genomes**, and **file conversion**.
These other applications are detailed in the [wiki](https://github.com/aast242/FACET/wiki).

## Database Module
Usage: `FACET database <subject.fasta> <query.fasta> [options]`

Acceptable Module Aliases: `database`, `db` 

This is the original way to run FACET. The program creates a BLASTn database from <subject.fasta> and queries the database with <query.fasta>, generating blast alignments. Using a database has the advantage of being able to use multiple cores while running BLAST, speeding up searches. BLAST databases are stored in the directory FACET is stored in. For more information on options, see the [wiki](https://github.com/aast242/FACET/wiki).


## Database-Free Module
Usage: `FACET db_free <subject.fasta> <query.fasta> [options]`

Acceptable Module Aliases: `db_free`, `dbf`, `database_free` 

FACET does not create a BLASTn database, but uses <subject.fasta> as the subect and queries it using <query.fasta> to generate blast alignments. This module has the advantage of not creating database files, but can be undesirably slow for larger jobs (i.e., a genome self-comparison). For more information on options, see the [wiki](https://github.com/aast242/FACET/wiki).


## Outfile Module
Usage: `FACET outfile <blastn.out> <subject.fasta> <query.fasta> [options]`

Acceptable Module Aliases: `outfile`

Uses an output generated by either BLASTn or FACET to perform cleaning or file conversion. The output needs to begin with FACET's outformat OR use BLAST's "-outfmt 6" for this module to operate correctly. Subject and query fasta files used to generate the BLASTn output are required for file outputs. For more information on options, see the [wiki](https://github.com/aast242/FACET/wiki).


## Masker Module
Usage: `FACET masker <genome.fasta> [options]`

Acceptable Module Aliases: `masker` 

Performs a genome self-comparison using a temporary BLAST database, does not perform any cleaning, and uses alignment information to mask repetitive regions in <genome.fasta>. For more information on options, see the [wiki](https://github.com/aast242/FACET/wiki).
